doc/designs/fips_indicator.md

```diff
@@ -0,0 +1,291 @@
+AWS/google have added a table in their security policy called
+‘Non-Approved Algorithms not allowed in the approved mode of operation’.
+An example is RSA with a keysize of < 2048 bits (which has been enforced by [4]).
+
+### Legacy Support
+
+Due to key transitions [4] we may have some legacy algorithms that are in a state of only being approved for processing (verification, decryption, validation), and not for protection (signing, encrypting, keygen).
+For example DSA.
+
+The options are:
+- Completely remove the algorithm from the FIPS provider.
+  This is simple but means older applications can no longer process existing data, which is not ideal.
+- Allow the algorithm but make it not approved with an context specific indicator.
+It is safer to make the protection operations fail rather than use an indicator.
+The processing operation for DSA would set the indicator to approved.
+
+
+### Security Checks
+
+OpenSSL currently defines configurable FIPS options.
+These options are supplied via the FIPS configuration file - which is normally setup via fipsinstall.
+
+- FIPS_FEATURE_CHECK(FIPS_security_check_enabled, fips_security_checks, 1)
+- FIPS_FEATURE_CHECK(FIPS_tls_prf_ems_check, fips_tls1_prf_ems_check, 0)
+- FIPS_FEATURE_CHECK(FIPS_restricted_drbg_digests_enabled, 0)
+- OSSL_PROV_FIPS_PARAM_CONDITIONAL_ERRORS selftest_params.conditional_error_check
+
+The following functions are available in providers/common/security_check.c.
+
+- ossl_rsa_check_key()
+- ossl_ec_check_key()
+- ossl_dsa_check_key()
+- ossl_dh_check_key()
+- ossl_digest_get_approved_nid_with_sha1()
+- ossl_digest_is_allowed()
+
+Anywhere where these functions are called an indicator MAY be required.
+Because these options are available I do not think it is sufficient to
+document this in the security policy.
+
+Each of these functions contains code of the following form:
+~~~
+#if !defined(OPENSSL_NO_FIPS_SECURITYCHECKS)
+    if (ossl_securitycheck_enabled(ctx)) {
+      // Do some checks, and maybe return 0 for a hard failure
+      ...
+    }
+}
+~~~
+
+OPENSSL_NO_FIPS_SECURITYCHECKS is also a configurable option
+If the security checks are not enabled then it is unapproved?
+
+
+## Implementation options
+
+The above requirements indicate 2 options.
+
+### Option 1
+- Dont allow ANY non approved algorithms and then a indicator is not required.
+
+#### Pros
+- Simple
+
+#### Cons
+- Problematic since we already have configurable options that are used for security checks etc.
+- We would need to return errors anywhere where an algorithm is not approved, which would cause
+  compatability issues
+
+### Preferred Option
+- Add an indicator everywhere that it is required.
+
+#### Pros
+- Flexible solution
+#### Cons
+- Requires a lot more effort to add the indicator to all the required places.
+
+
+Note that in order for a service to be ‘fips approved’ the following requirments would need to be met.
+- Any algorithms are fetched using “fips=yes”
```

Comment:
IMO. this isn't required.  They **must**, however, be fetched from a FIPS approved provider.

Reply:
In that case the ones that have fips=no need an indicator.

Reply:
It was my understanding that the FIPS provider (i.e. fips.so) defined the module boundary. Therefore any algorithms outside the FIPS provider would simply not be part of the FIPS module (and thus not listed in the Security Policy). Does the "fips=yes" property not enforce algorithms to be fetched through the FIPS provider?

Reply:
That is the definition of an implicit indicator.
I want the indicator to apply to any provider i.e. retrieving the getter on any ctx will return either yes - its FIPS approved OR no its not.

Reply:
Well... there could be helper functions but if implemented as OSSL_PARAM, it should be that a param is set to true if FIPS approved/allowed, otherwise - not present or false -> not approved.

I think it would be bad to make this unclear - i.e. the param value not returned means something _undefined_.